### ISSUE TITLE
Semantic Segmentation testing on Jetson TX1

### ISSUE THREAD

**Issue by akshaykulkarni07** [akshaykvnit/autonomous-delivery-robot#1]
We need to test the semantic segmentation model (of Keras/TF) on Nvidia Jetson TX1, calculate the FPS and identify whether we require additional hardware for this computation.

**Comment by akshaykulkarni07** [comment]
This task requires installation of TensorFlow on Jetson. Currently, there is some issue with the setup of Jetson (the OS is not on SD card), so enough space is not available for TF installation. Workaround is provided [here](https://www.jetsonhacks.com/2017/01/26/run-jetson-tx1-sd-card/).

[Jetson Zoo](https://elinux.org/Jetson_Zoo) provides useful installation procedures for common softwares.

**Comment by akshaykulkarni07** [comment]
The installation of Jetpack 3.1 has CUDA 8.0, while TensorFlow requires CUDA 9.0. So, we tried the installation of both Jetpack 3.3 and 3.2 which has CUDA 9.0, but there were some issues during the installation, so it needs to be repeated. This will be done after coming back from home (have to work on FYP presentation now).

**Comment by akshaykulkarni07** [comment]
Take a look at testing videos processed on the GTX 1080Ti at [this link](https://drive.google.com/drive/folders/1ZWK42bZJJxiEu8831j1k6gu-Vfkt8kwf?usp=sharing).

**Comment by akshaykvnit** [comment]
Since, older versions (16.04) weren't working, we tried using JetPack 4.2.1 (18.04). However, the shifting to SD card workaround mentioned earlier (from [here](https://www.jetsonhacks.com/2017/01/26/run-jetson-tx1-sd-card/)) doesn't work directly in 18.04. So, we'll try to install the older JetPack, then shift to SD card and then install PyTorch. Hopefully, this will work and we can shift our code to PyTorch. 
Worst case scenario, we'll have to use a Raspberry Pi 3.

**Comment by akshaykulkarni07** [comment]
Shifted 18.04 Ubuntu to SD Card (using same link mentioned above). Installed PyTorch 1.3.0 using [this link](https://elinux.org/Jetson_Zoo#PyTorch_.28Caffe2.29) for Python 2 because ROS supports only Python2. torchvision library is installed using [this link](https://devtalk.nvidia.com/default/topic/1049071/jetson-nano/pytorch-for-jetson-nano-version-1-3-0-now-available/).

Will shift all code to PyTorch (and Python2) now. This will take at least one week.

**Comment by akshaykulkarni07** [comment]
Shifted all code (of ENet) to PyTorch (Python2). Tested the code on Jetson TX1, got around 8 FPS for inputs of 640x320. However, network isn't trained fully yet. Need to do this, but the code works on Jetson, so this task is done. 

Note: We had to [add a swap file](https://www.jetsonhacks.com/2016/12/21/jetson-tx1-swap-file-and-development-preparation/) (of 8 GB) since Jetson TX1 has a shared RAM and VRAM of total 4 GB. Otherwise, a low memory warning and subsequent killing of process occurs on running the model.

**Comment by navidpanchi** [comment]
You have predicted on a single image at a time, you  should use batch prediction. That could give many more fps than this.

**Comment by akshaykulkarni07** [comment]
What do we do with those multiple predictions? I think we need only the latest one. Unless we are thinking of averaging the predictions of the last few images. It's probably not a good idea because the scene will have changed. 
Also, even if we do batch prediction, it will still require at least the time for a single image (0.125 seconds according to 8 FPS). And we won't be predicting images continuously anyway, so there's no advantage in batch processing.  
Any further comments @navidpanchi?
